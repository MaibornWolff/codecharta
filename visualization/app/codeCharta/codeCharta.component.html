<tool-bar-component></tool-bar-component>

<file-extension-bar-component></file-extension-bar-component>
<div class="ribbon-bar-wrapper">
	<ribbon-bar-component class="ribbon-bar-component"></ribbon-bar-component>
</div>

<code-map-component></code-map-component>

<legend-panel-component></legend-panel-component>

<loading-gif-file-component></loading-gif-file-component>

<div id="mw-logo">
	<div class="logo">
		<a href="https://www.maibornwolff.de" target="_blank"><img src="assets/logo.png" alt="MaibornWolff"/></a>
		<h2>
			<span>CodeCharta {{ $ctrl._viewModel.version }}</span>
		</h2>
	</div>
</div>

<node-context-menu-component></node-context-menu-component>
