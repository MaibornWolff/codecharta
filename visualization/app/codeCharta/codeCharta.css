*, *:before, *:after {
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
}

html, body {
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    font-size: 1em;
    cursor: default;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#codeMap {
    position: absolute;
    z-index: 0;
    top: 50px;
    left: 0;
    width: 100%;
    height: 100%;
}

#hud {
    position: absolute;
    z-index: 10;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    max-width: none !important;
    pointer-events: none;
    padding-left:10px;
    padding-right:10px;
}

revision-chooser-component, legend-panel-component, file-chooser-panel-directive, settings-panel-component, .mwlink {
    pointer-events: auto;
}

#title {
    text-transform: uppercase;
    font-size: 0.9em;
    line-height: 1.4em;
    float: right;
    margin: 0.5em 1.6em 0 0;
    color: #5A585A;
    font-family: Roboto, sans-serif;
    font-weight: 300;
}

#title:first-child {
    font-size: 1.82em;
}

#title:last-child {
    font-size: 1.5em;
}

.collapsible > li {
    background-color: #ffffff !important;
}

.col-backdrop {
    background-color: rgba(0,0,0,0.1);
}

.bottom-row {
    position: fixed;
    bottom: 0;
    width:100%;
}

.mwlink > img{
    width: 100%;
    max-width: 4em;
    float:right;
}

.chip {
    background-color: #efefef;
}

.row, .col {
    min-height: 0px !important;
}


/*

#actionButtons {
    position: fixed;
    bottom: 2em;
    right: 2em;
    z-index: 1000;
}

#settingsPanel {
    position: fixed;
    right: 0;
    width: 0px;
    height: 100%;
    background: #fff;
    z-index: 1000;
    box-shadow: 0 0 10px #000;
}

#detailWindow {
    position: fixed;
    z-index: 999;
    bottom: 0.5em;
    left: 0.5em;
    color:#5A585A;
    font-size: 0.8em;
}

#legendButton {
    position: absolute;
    top: 11.5em;
    left: -2.9em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#legendPanel {
    position: fixed;
    top: 9em;
    left: -500px;
}

#revisionButton {
    position: absolute;
    top: 21em;
    left: -3.5em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#mapButton {
    position: absolute;
    top: 21em;
    left: -2.3em;
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}

#revisionChooser {
    position: fixed;
    top: 18.5em;
    left: -500px;
}

.rzslider .rz-pointer, .rzslider .rz-selection {
    background: #26a69a !important;
}

.rzslider .rz-bar {
    background: rgb(196,196,196);
}*/