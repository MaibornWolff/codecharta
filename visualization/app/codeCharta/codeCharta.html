<div id="hud" class="container">

    <div class="row">

        <div class="col l3 m6 s12 col-backdrop">
            <legend-panel-component></legend-panel-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <revision-chooser-component></revision-chooser-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <file-chooser-panel-directive></file-chooser-panel-directive>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <settings-panel-component></settings-panel-component>
        </div>

    </div>

    <div class="row bottom-row">

        <div class="col l3 m6 s12 col-backdrop">
            <detail-panel-component></detail-panel-component>
        </div>

    </div>

    <div id="titlebox">
        <a class="mwlink" href="https://www.maibornwolff.de" target="_blank"><img src="./logo.png" alt="MaibornWolff"></a>
        <h1 id="title"><span>CodeCharta {{$ctrl.pkg.version}}</span></h1>
    </div>

</div>

<code-map-component></code-map-component>