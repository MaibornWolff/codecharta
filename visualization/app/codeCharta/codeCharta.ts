"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


