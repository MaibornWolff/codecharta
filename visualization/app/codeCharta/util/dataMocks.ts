import { CCFile, CodeMapNode, Edge, MetricData, Node, Settings } from "../codeCharta.model"
import { CodeMapBuilding } from "../ui/codeMap/rendering/codeMapBuilding"
import * as THREE from "three"

export const VALID_NODE: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	children: [
		{
			name: "big leaf",
			type: "File",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			children: [
				{
					name: "small leaf",
					type: "File",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				}
			]
		}
	]
}

export const VALID_NODE_WITH_PATH: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	path: "/root",
	children: [
		{
			name: "big leaf",
			type: "File",
			path: "/root/big leaf",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			path: "/root/Parent Leaf",
			children: [
				{
					name: "small leaf",
					type: "File",
					path: "/root/Parent Leaf/small leaf",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					path: "/root/Parent Leaf/other small leaf",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				},
				{
					name: "empty folder",
					type: "Folder",
					path: "/root/Parent Leaf/empty folder",
					attributes: {},
					children: []
				}
			]
		}
	]
}

export const VALID_EDGES: Edge[] = [
	{
		fromNodeName: "/root/big leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	},
	{
		fromNodeName: "/root/sample1 only leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	}
]

export const VALID_EDGE: Edge = {
	fromNodeName: "/root/big leaf",
	toNodeName: "/root/Parent Leaf/small leaf",
	attributes: {
		pairingRate: 89,
		avgCommits: 34
	}
}

export const TEST_FILE_CONTENT = {
	fileName: "noFileName",
	projectName: "Sample Map",
	apiVersion: "1.1",
	nodes: [VALID_NODE]
}

export const TEST_FILE_DATA: CCFile = {
	fileMeta: {
		fileName: "file",
		projectName: "Sample Map",
		apiVersion: "1.1"
	},
	map: VALID_NODE,
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_WITH_PATHS: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		path: "/root",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				path: "/root/big leaf",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				path: "/root/Parent Leaf",
				children: [
					{
						name: "small leaf",
						type: "File",
						path: "/root/Parent Leaf/small leaf",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						path: "/root/Parent Leaf/other small leaf",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					},
					{
						name: "empty folder",
						type: "Folder",
						path: "/root/Parent Leaf/empty folder",
						attributes: {},
						children: []
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_A: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_B: CCFile = {
	fileMeta: {
		fileName: "fileB",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 20, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "additional leaf",
				type: "File",
				attributes: { rloc: 10, functions: 11, mcc: 5 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100, more: 20 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000 }
					},
					{
						name: "big leaf",
						type: "File",
						attributes: { rloc: 20, functions: 10, mcc: 1 },
						link: "http://www.google.de"
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_DATA_DOWNLOADED = {
	apiVersion: "1.1",
	attributeTypes: {},
	blacklist: [],
	edges: [
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/big leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		},
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/sample1 only leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		}
	],
	fileName: "file_2018-12-14_9-39.cc.json",
	nodes: [
		{
			attributes: {},
			children: [
				{
					attributes: { Functions: 10, MCC: 1, RLOC: 100 },
					link: "http://www.google.de",
					name: "big leaf",
					type: "File"
				},
				{
					attributes: {},
					children: [
						{
							attributes: { Functions: 100, MCC: 100, RLOC: 30 },
							name: "small leaf",
							type: "File"
						},
						{
							attributes: { Functions: 1000, MCC: 10, RLOC: 70 },
							name: "other small leaf",
							type: "File"
						}
					],
					name: "Parent Leaf",
					type: "Folder"
				}
			],
			name: "root",
			type: "Folder"
		}
	],
	projectName: "Sample Map"
}

export const SETTINGS: Settings = {
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	dynamicSettings: {
		areaMetric: "rloc",
		heightMetric: "mcc",
		colorMetric: "mcc",
		focusedNodePath: "/root",
		searchedNodePaths: [],
		searchPattern: "",
		margin: 48,
		neutralColorRange: { flipped: false, from: 19, to: 67 }
	},
	appSettings: {
		amountOfTopLabels: 31,
		scaling: new THREE.Vector3(1, 1.8, 1),
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		maximizeDetailPanel: false,
		invertHeight: true,
		dynamicMargin: true,
		isWhiteBackground: false,
		whiteColorBuildings: true,
		mapColors: {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"]
		}
	},
	treeMapSettings: { mapSize: 500 }
}

export const DEFAULT_SETTINGS: Settings = {
	appSettings: {
		amountOfTopLabels: 1,
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		dynamicMargin: true,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		invertHeight: false,
		isWhiteBackground: false,
		mapColors: {
			angularGreen: "#00BFA5",
			base: "#666666",
			defaultC: "#89ACB4",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"],
			negative: "#820E0E",
			negativeDelta: "#ff0E0E",
			neutral: "#ddcc00",
			positive: "#69AE40",
			positiveDelta: "#69FF40",
			selected: "#EB8319"
		},
		maximizeDetailPanel: false,
		scaling: new THREE.Vector3(1, 1, 1),
		whiteColorBuildings: false
	},
	dynamicSettings: {
		areaMetric: null,
		colorMetric: null,
		focusedNodePath: "",
		heightMetric: null,
		margin: null,
		neutralColorRange: {
			flipped: false,
			from: null,
			to: null
		},
		searchPattern: "",
		searchedNodePaths: []
	},
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	treeMapSettings: { mapSize: 500 }
}

export const TEST_NODE_ROOT: Node = {
	name: "root",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 10,
	visible: true,
	path: "root",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0x000000",
	flat: false
}

export const TEST_NODE_LEAF: Node = {
	name: "root/big leaf",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 20,
	visible: true,
	path: "root/big leaf",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0xFFFFFF",
	flat: false
}

export const CODE_MAP_BUILDING: CodeMapBuilding = new CodeMapBuilding(
	1,
	new THREE.Box3(),
	TEST_NODE_ROOT,
	DEFAULT_SETTINGS.appSettings.mapColors.neutral
)

export const METRIC_DATA: MetricData[] = [
	{ name: "mcc", maxValue: 1, availableInVisibleMaps: true },
	{ name: "rloc", maxValue: 2, availableInVisibleMaps: true }
]
