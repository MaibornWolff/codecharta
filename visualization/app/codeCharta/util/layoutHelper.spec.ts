import { CodeMapNode, NodeType, State, EdgeVisibility, BlacklistType } from "../codeCharta.model"
import { LayoutNode, LayoutHelper } from "./layoutHelper"
import Rectangle from "./algorithm/rectangle"
import Point from "./algorithm/point"
import { STATE, CODE_MAP_BUILDING } from "./dataMocks"

describe("layoutHelper", () => {
	describe("build node", () => {
		let codeMapNode: CodeMapNode
		let layoutNode: LayoutNode
		let state

		let heightScale = 1
		let maxHeight = 2000
		let isDeltaState = false

		beforeEach(() => {
			codeMapNode = {
				name: "Anode",
				path: "/root/Anode",
				type: NodeType.FILE,
				attributes: { theHeight: 100 }
			} as CodeMapNode

			layoutNode = {
				data: codeMapNode,
				value: 42,
				rect: new Rectangle(new Point(0, 0), 400, 400),
				zOffset: 0
			}

			state = STATE
			state.treeMap.mapSize = 1
			state.dynamicSettings.margin = 15
			state.dynamicSettings.heightMetric = "theHeight"
			state.appSettings.invertHeight = false
		})

		function buildNode() {
			return LayoutHelper.buildNodeFrom(layoutNode, heightScale, maxHeight, state, isDeltaState)
		}

		it("minimal", () => {
			expect(buildNode()).toMatchSnapshot()
		})

		it("invertHeight", () => {
			state.appSettings.invertHeight = true
			expect(buildNode()).toMatchSnapshot()
		})

		it("deltas", () => {
			layoutNode.data.deltas = {}
			state.dynamicSettings.heightMetric = "theHeight"
			layoutNode.data.deltas[state.dynamicSettings.heightMetric] = 33
			expect(buildNode()).toMatchSnapshot()
			layoutNode.data.deltas = undefined
		})

		it("given negative deltas the resulting heightDelta also should be negative", () => {
			layoutNode.data.deltas = {}
			layoutNode.data.deltas[state.dynamicSettings.heightMetric] = -33
			expect(buildNode().heightDelta).toBe(-33)
			layoutNode.data.deltas = undefined
		})

		it("should set lowest possible height caused by other visible edge pairs", () => {
			state.fileSettings.edges = [
				{
					fromNodeName: "/root/AnotherNode1",
					toNodeName: "/root/AnotherNode2",
					attributes: {
						pairingRate: 33,
						avgCommits: 12
					},
					visible: true
				}
			]
			expect(buildNode()).toMatchSnapshot()
		})

		it("should set markingColor according to markedPackages", () => {
			const color = "#FF0000"
			state.fileSettings.markedPackages = [
				{
					path: "/root/Anode",
					color: color,
					attributes: {}
				}
			]
			expect(buildNode().markingColor).toEqual(color)
		})

		it("should set no markingColor according to markedPackages", () => {
			const color = "#FF0000"
			state.fileSettings.markedPackages = [
				{
					path: "/root/AnotherNode",
					color: color,
					attributes: {}
				}
			]
			expect(buildNode().markingColor).toEqual(null)
		})
	})

	describe("count nodes", () => {
		it("root only should be 1", () => {
			const root = {}
			expect(LayoutHelper.countNodes(root)).toBe(1)
		})

		it("root plus child should be 2", () => {
			const root = { children: [{}] }
			expect(LayoutHelper.countNodes(root)).toBe(2)
		})

		it("root plus child in child should be 3", () => {
			const root = { children: [{ children: [{}] }] }
			expect(LayoutHelper.countNodes(root)).toBe(3)
		})

		it("root plus two children should be 3", () => {
			const root = { children: [{}, {}] }
			expect(LayoutHelper.countNodes(root)).toBe(3)
		})
	})

	describe("isNodeToBeFlat", () => {
		let codeMapNode: CodeMapNode
		let layoutNode: LayoutNode
		let state: State

		beforeEach(() => {
			codeMapNode = {
				name: "Anode",
				path: "/root/Anode",
				type: NodeType.FILE,
				attributes: {},
				edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } }
			}

			layoutNode = {
				data: codeMapNode,
				value: 42,
				rect: new Rectangle(new Point(0, 0), 400, 400),
				zOffset: 0
			}

			state = STATE
			state.treeMap.mapSize = 1
			state.dynamicSettings.margin = 15
		})

		it("should not be a flat node when no visibleEdges", () => {
			state.fileSettings.edges = []
			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeFalsy()
		})

		it("should be a flat node when other edges are visible", () => {
			state.appSettings.showOnlyBuildingsWithEdges = true
			state.fileSettings.edges = [
				{
					fromNodeName: "/root/anotherNode",
					toNodeName: "/root/anotherNode2",
					attributes: {},
					visible: EdgeVisibility.both
				}
			]
			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeTruthy()
		})

		it("should not be a flat node when it contains edges", () => {
			state.fileSettings.edges = [
				{
					fromNodeName: "/root/Anode",
					toNodeName: "/root/anotherNode",
					attributes: {}
				}
			]
			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeFalsy()
		})

		it("should not be a flat node, because its searched for", () => {
			state.dynamicSettings.searchedNodePaths = ["/root/Anode"]
			state.dynamicSettings.searchPattern = "Anode"
			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeFalsy()
		})

		it("should be a flat node, because other nodes are searched for", () => {
			state.dynamicSettings.searchedNodePaths = ["/root/anotherNode", "/root/anotherNode2"]
			state.dynamicSettings.searchPattern = "Anode"
			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeTruthy()
		})

		it("should not be a flat node when searchPattern is empty", () => {
			state.dynamicSettings.searchedNodePaths = ["/root/anotherNode", "/root/anotherNode2"]
			state.dynamicSettings.searchPattern = ""
			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeFalsy()
		})

		it("should be flat if node is flattened in blacklist", () => {
			state.fileSettings.blacklist = [{ path: "*Anode", type: BlacklistType.flatten }]

			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeTruthy()
		})

		it("should not be flat if node is not blacklisted", () => {
			state.fileSettings.blacklist = []

			expect(LayoutHelper["isNodeToBeFlat"](layoutNode, state)).toBeFalsy()
		})
	})

	describe("getBuildingColor", () => {
		let node: CodeMapNode
		let state: State

		beforeEach(() => {
			node = {
				name: "Anode",
				path: "/root/Anode",
				type: NodeType.FILE,
				attributes: {}
			} as CodeMapNode

			node.attributes = { validMetircName: 0 }

			state = STATE
			state.appSettings.invertColorRange = false
			state.appSettings.whiteColorBuildings = false
			state.dynamicSettings.colorRange.from = 5
			state.dynamicSettings.colorRange.to = 10
			state.dynamicSettings.colorMetric = "validMetircName"
		})

		it("creates grey building for undefined colorMetric", () => {
			state.dynamicSettings.colorMetric = "invalid"
			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)
			expect(buildingColor).toBe(state.appSettings.mapColors.base)
		})

		it("creates flat colored building", () => {
			const flattend = true

			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, flattend)

			expect(buildingColor).toBe(state.appSettings.mapColors.flat)
		})

		it("creates green colored building colorMetricValue < colorRangeFrom", () => {
			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)

			expect(buildingColor).toBe(state.appSettings.mapColors.positive)
		})

		it("creates white colored building colorMetricValue < colorRangeFrom", () => {
			state.appSettings.whiteColorBuildings = true

			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)

			expect(buildingColor).toBe(state.appSettings.mapColors.lightGrey)
		})

		it("creates red colored building colorMetricValue < colorRangeFrom with inverted range", () => {
			state.appSettings.invertColorRange = true

			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)

			expect(buildingColor).toBe(state.appSettings.mapColors.negative)
		})

		it("creates red colored building colorMetricValue > colorRangeFrom", () => {
			node.attributes = { validMetircName: 12 }

			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)

			expect(buildingColor).toBe(state.appSettings.mapColors.negative)
		})

		it("creates green colored building colorMetricValue > colorRangeFrom with inverted range", () => {
			state.appSettings.invertColorRange = true
			node.attributes = { validMetircName: 12 }

			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)

			expect(buildingColor).toBe(state.appSettings.mapColors.positive)
		})

		it("creates white colored building colorMetricValue > colorRangeFrom with inverted range", () => {
			state.appSettings.invertColorRange = true
			state.appSettings.whiteColorBuildings = true
			node.attributes = { validMetircName: 12 }

			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)

			expect(buildingColor).toBe(state.appSettings.mapColors.lightGrey)
		})

		it("creates yellow colored building", () => {
			node.attributes = { validMetircName: 7 }
			const buildingColor = LayoutHelper["getBuildingColor"](node, state, false, false)
			expect(buildingColor).toBe(state.appSettings.mapColors.neutral)
		})
	})

	describe("buildingArrayToMap", () => {
		it("should convert a array of buildings to a map", () => {
			const result = LayoutHelper.buildingArrayToMap([CODE_MAP_BUILDING])

			expect(result.get(CODE_MAP_BUILDING.id)).toEqual(CODE_MAP_BUILDING)
		})
	})

	describe("isNodeLeaf", () => {
		let innerNode: CodeMapNode
		let leafNode: CodeMapNode

		beforeEach(() => {
			leafNode = {
				name: "Anode",
				path: "/root/Anode",
				type: NodeType.FILE,
				attributes: {},
				edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } }
			} as CodeMapNode

			innerNode = {
				name: "root",
				path: "/root",
				type: NodeType.FOLDER,
				attributes: {},
				children: [leafNode],
				edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } }
			}

			it("should be leaf", () => {
				const isFile = LayoutHelper.isNodeLeaf(leafNode)
				expect(isFile).toBe(true)
			})

			it("should not be leaf", () => {
				const isFile = LayoutHelper.isNodeLeaf(innerNode)
				expect(isFile).toBe(false)
			})
		})
	})

	describe("calculateSize", () => {
		let innerNode: CodeMapNode
		let leafNode: CodeMapNode

		beforeEach(() => {
			leafNode = {
				name: "Anode",
				path: "/root/Anode",
				type: NodeType.FILE,
				attributes: { rloc: 100 },
				edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } }
			} as CodeMapNode

			innerNode = {
				name: "root",
				path: "/root",
				type: NodeType.FOLDER,
				attributes: {},
				children: [leafNode, leafNode],
				edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } }
			} as CodeMapNode

			it("should be the node's attribute", () => {
				const size = LayoutHelper.calculateSize(leafNode, "rloc")
				expect(size).toBe(leafNode.attributes["rloc"])
			})

			it("should be the sum of all its childrens' summed attributes", () => {
				const isFile = LayoutHelper.isNodeLeaf(innerNode)
				expect(isFile).toBe(leafNode.children.reduce((a, b) => a + b.attributes["rloc"], 0))
			})
		})
	})

	describe("mergeDirectories", () => {
		let innerNode: CodeMapNode
		let leafNode: CodeMapNode

		beforeEach(() => {
			leafNode = {
				name: "Anode",
				path: "/root/Anode",
				type: NodeType.FILE,
				attributes: { rloc: 100 },
				edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } }
			} as CodeMapNode

			innerNode = {
				name: "root",
				path: "/root",
				type: NodeType.FOLDER,
				attributes: {},
				children: [leafNode],
				edgeAttributes: { pairingRate: { incoming: 42, outgoing: 23 } }
			} as CodeMapNode

			it("should be merged directory names", () => {
				const node = LayoutHelper.mergeDirectories(innerNode, "rloc")
				expect(node.name).toBe(innerNode.name + "/" + leafNode.name)
			})

			it("should not be merged names", () => {
				const node = LayoutHelper.mergeDirectories(innerNode, "rloc")
				expect(node.name).toBe(innerNode.name)
			})
		})
	})
})
