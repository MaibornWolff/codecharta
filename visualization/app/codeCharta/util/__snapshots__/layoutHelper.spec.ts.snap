// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layoutHelper build node deltas 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": Object {
    "theHeight": 33,
  },
  "depth": 0,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 100,
  "heightDelta": 33,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 100,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 100,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`layoutHelper build node invertHeight 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": undefined,
  "depth": 0,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 1900,
  "heightDelta": 0,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 1900,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 1900,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`layoutHelper build node minimal 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": undefined,
  "depth": 0,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 100,
  "heightDelta": 0,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 100,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 100,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;

exports[`layoutHelper build node should set lowest possible height caused by other visible edge pairs 1`] = `
Object {
  "attributes": Object {
    "theHeight": 100,
  },
  "color": "#666666",
  "deltas": undefined,
  "depth": 0,
  "edgeAttributes": undefined,
  "flat": false,
  "height": 100,
  "heightDelta": 0,
  "incomingEdgePoint": Vector3 {
    "x": 199,
    "y": 100,
    "z": 99,
  },
  "isLeaf": true,
  "length": 400,
  "link": undefined,
  "markingColor": null,
  "name": "Anode",
  "outgoingEdgePoint": Vector3 {
    "x": 199,
    "y": 100,
    "z": 299,
  },
  "path": "/root/Anode",
  "visible": undefined,
  "width": 400,
  "x0": 0,
  "y0": 0,
  "z0": 0,
}
`;
